discoverd2/server/dns_test.go

```diff
@@ -0,0 +1,552 @@
+	}
+	v6LeaderQs := map[uint16][]testAddr{
+		dns.TypeA:    nil,
+		dns.TypeAAAA: v6v4Addrs[:1],
+		dns.TypeANY:  v6v4Addrs[:1],
+		dns.TypeSRV:  v6v4Addrs[:1],
+		dns.TypeSOA:  nil,
+		dns.TypeTXT:  nil,
+	}
+
+	tests := []test{
+		{
+			name:   "non-existent service NXDOMAIN",
+			domain: "b.discoverd.",
+			qs:     emptyAll,
+		},
+		{
+			name:   "non-existent service RFC2782 NXDOMAIN",
+			domain: "_b._tcp.discoverd.",
+			qs:     emptyAll,
+		},
+		{
+			name:   "non-existent instance with existing service NXDOMAIN",
+			domain: "foo.a._i.discoverd.",
+			qs:     emptyAll,
+		},
+		{
+			name:   "non-existent instance with non-existent service NXDOMAIN",
+			domain: "foo.b._i.discoverd.",
+			qs:     emptyAll,
+		},
+		{
+			name:   "bad domain NXDOMAIN",
+			domain: "asdf.b.discoverd.",
+			qs:     emptyAll,
+		},
+		{
+			name:   "leader NXDOMAIN",
+			domain: "leader.b.discoverd.",
+			qs:     emptyAll,
+		},
+		{
+			name:   "leader with non-existent service NXDOMAIN",
+			domain: "leader.b.discoverd.",
+			qs:     emptyAll,
+		},
+		{
+			name:   "empty service",
+			domain: "a.discoverd.",
+			qs:     emptyAll,
+		},
+		{
+			name:   "empty 2782",
+			domain: "_a._tcp.discoverd.",
+			qs:     emptyAll,
+		},
+		{
+			name:   "2782 no proto matches",
+			domain: "_a._foo.discoverd.",
+			data:   simpleData,
+			qs:     emptyAll,
+		},
+		{
+			name:   "2782 full proto matches",
+			domain: "_a._tcp.discoverd.",
+			data:   simpleData,
+			qs:     simpleQs,
+		},
+		{
+			name:   "2782 some proto matches",
+			domain: "_a._udp.discoverd.",
+			data:   mixedData,
+			qs:     udpFiltered,
+		},
+		{
+			name:   "service mixed protocols",
+			domain: "a.discoverd.",
+			data:   mixedData,
+			qs:     mixedQs,
+		},
+		{
+			name:   "service mixed v4/v6",
+			domain: "a.discoverd.",
+			data:   v6v4Data,
+			qs:     v6v4Qs,
+		},
+		{
+			name:   "2782 mixed v4/v6",
+			domain: "_a._tcp.discoverd.",
+			data:   v6v4Data,
+			qs:     v6v4Qs,
+		},
+		{
+			name:   "instance",
+			domain: fmt.Sprintf("%s.a._i.discoverd.", simpleData[0].ID),
+			data:   simpleData,
+			qs:     instanceQs,
+		},
+		{
+			name:   "v6 instance",
+			domain: fmt.Sprintf("%s.a._i.discoverd.", v6v4Data[0].ID),
+			data:   v6v4Data,
+			qs:     v6instanceQs,
+		},
+		{
+			name:   "service udp limit",
+			domain: "a.discoverd.",
+			data:   longData,
+			qs:     longQs,
+		},
+		{
+			name:   "2782 udp limit",
+			domain: "_a._tcp.discoverd.",
+			data:   longData,
+			qs:     longQs,
+		},
+		{
+			name:   "service duplicate IPs",
+			domain: "a.discoverd.",
+			data:   dupeData,
+			qs:     dupeQs,
+		},
+		{
+			name:   "2782 duplicate IPs",
+			domain: "_a._tcp.discoverd.",
+			data:   dupeData,
+			qs:     dupeQs,
+		},
+		{
+			name:   "leader",
+			domain: "leader.a.discoverd.",
+			data:   simpleData,
+			qs:     leaderQs,
+		},
+		{
+			name:   "v6 leader",
+			domain: "leader.a.discoverd.",
+			data:   v6v4Data[:1],
+			qs:     v6LeaderQs,
+		},
```

Comment:
Do any of these test datasets hit the recursors?

(I guess that's a hard area to reach... if we can't swallow the network dependency, maybe mocked recursors?)